src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/****************************************************************************
+ * @file
+ * @brief Implementation for the Administrator Commissioning Cluster
+ ***************************************************************************/
+
+#include <app/CommandHandler.h>
+#include <app/server/Server.h>
+#include <app/util/af.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip;
+
+bool emberAfAdministratorCommissioningClusterOpenCommissioningWindowCallback(chip::EndpointId endpoint,
+                                                                             app::CommandHandler * commandObj,
+                                                                             uint16_t commissioningTimeout, ByteSpan pakeVerifier,
+                                                                             uint16_t discriminator, uint32_t iterations,
+                                                                             ByteSpan salt, uint16_t passcodeID)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    PASEVerifier verifier;
+    ChipLogProgress(Zcl, "Received command to open commissioning window");
+    VerifyOrExit(!IsPairingWindowOpen(), status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(sizeof(verifier) == pakeVerifier.size(), status = EMBER_ZCL_STATUS_FAILURE);
```

Comment:
Please add check that:
1- Iterations is within expected range (10000 .. 100000)
2- salt has correct length (16..32 bytes).

Per spec in Crypto primitives:

```
[width="100%", cols="1a", options="header"]
|===
|Mapping (Version 1)
|`int CHIP_CRYPTO_PBKDF_ITERATIONS_MIN := 1000`
|`int CHIP_CRYPTO_PBKDF_ITERATIONS_MAX := 100000`
|
----
CHIP_Crypto_PBKDF(input, salt, iterations, len) :=
     bit[len] PBKDF2(P := input, S := salt, C := iterations, kLen := len)
----
`PBKDF2()` SHALL be the HMAC-based PBKDF function with `CHIP_Crypto_HMAC(key := P, message := U[j-1])` as the auxiliary function `HMAC` as defined in Section 5.3 of <<NIST800132>>; it returns a bit array of `len` bits.
|
----
CHIP_Crypto_PBKDFParameterSet => STRUCTURE [ tag-order ]
{
     iterations [1] : UNSIGNED INTEGER [ range 32-bits ],
     salt [2] : OCTET STRING [ length 16..32 ],
}
----
* `*iterations*`: An integer value specifying the number of PBKDF2 iterations: `CHIP_CRYPTO_PBKDF_ITERATIONS_MIN` stem:[<=] `iterations` stem:[<=] `CHIP_CRYPTO_PBKDF_ITERATIONS_MAX`.
* `*salt*`: A random value per device of at least 16 bytes and at most 32 bytes used as the PBKDF2 salt.
|===
```